### PR TITLE
 added more changes to handle the infinite mode better when not enoug…

### PR DESCRIPTION
**Pull request by abhinavdalal-iconnect**
…h children

### CODE REVIEW

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG ^ this pr only has my changes now

**Comment by YIZHUANG** [comment]
Can you give an example what this PR fixes?

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG my apologies for not replying earlier..

Changes in this PR address:
1) typo fixed => getInitialSlideInInfiniteMode
2) reset the currentSlide when number of children change between updates (my use case is that i have debounced search query that filters the children passed)
3) reset the transform to 0 when not enough children (missed from last PR where we said no clones even with infinite mode)
4) dont add transform for center mode if not enough children(even with infinite active)

**Comment by YIZHUANG** [comment]
Can you provide a reproduction for point 3?

**Comment by abhinavdalal-iconnect** [comment]
will get a working demo ready for point no. 3 by tomorrow. although i think it only happens on resize.. first render works as expected. But on resize to a larger screen; it would not realize not to use infinite mode transform when not enough children. try mobile slidesToShow = 3 and desktop slidesToShow = 5 and totalItems =3; start in mobile view and resize to desktop.. it should reproduce the issue no. 3.

I will get a working demo for this ready by tomorrow.

**Comment by abhinavdalal** [comment]
steps to demo issue no. 3:
1) clone `https://github.com/abhinavdalal/matrimony-demo.git`
2) git checkout demo_infinite_mode_error
3) npm i
4) npm start

5) in localhost:3000 add username 'shaadi' and password '123' and submit
6) reduce size of the window to ~500px
7) select 3 in the drop down; you should see 3 items in the carousel in infinite mode (works perfectly fine)
8) resize the window to full desktop; you will see that the carousel disappears.. this issue is due to the transform not updating to notEnoughChildren mode of ignoring infinite
9) hit refresh; works as expected once refreshed => should show only 3 items in non-infinite mode.

File: src/Carousel.tsx [hunk 1]
```diff
@@ -126,30 +126,33 @@ class Carousel extends React.Component<CarouselProps, CarouselInternalState> {
...
+      childrenArr.length < this.state.slidesToShow
+        ? 0
+        : this.state.currentSlide;
+    this.setState(
+      {
+        totalItems: clones.length,
+        currentSlide: forResizing && !resetCurrentSlide ? currentSlide : initialSlide,
+      },
+      () => {
+        this.correctItemsPosition(itemWidth || this.state.itemWidth);
```
Comment (YIZHUANG) [inline src/Carousel.tsx]: The changes in this function is not needed, as it shouldn't be called at all. ``` TotalItems``` will be default to ```this.props.children.length``` if not specified.

File: src/Carousel.tsx [hunk 2]
```diff
@@ -205,7 +208,7 @@ class Carousel extends React.Component<CarouselProps, CarouselInternalState> {
       this.isAnimationAllowed = false;
     }
     this.setState({
-      transform: -(itemWidth * this.state.currentSlide)
+      transform: (this.state.totalItems < this.state.slidesToShow) ? 0 : -(itemWidth * this.state.currentSlide),
```
Comment (YIZHUANG) [inline src/Carousel.tsx]: This change is also not needed as currentSlide will already be 0 if there's isnt enough children.

File: src/Carousel.tsx [hunk 3]
```diff
@@ -255,8 +258,12 @@ class Carousel extends React.Component<CarouselProps, CarouselInternalState> {
       this.autoPlay = setInterval(this.next, this.props.autoPlaySpeed);
     }
     if (this.props.infinite) {
-      // this is to quickly cancel the animation and move the items position to create the infinite effects.
-      this.correctClonesPosition({ domLoaded });
+      if (children.length !== this.props.children.length) {
+        this.setClones(this.state.slidesToShow, this.state.itemWidth, true, true );
```
Comment (YIZHUANG) [inline src/Carousel.tsx]: 3 args.

File: src/utils/common.ts [hunk 4]
```diff
@@ -55,7 +55,7 @@ function getTransformForCenterMode(
   state: CarouselInternalState,
   props: CarouselProps
 ) {
-  if (state.currentSlide === 0 && !props.infinite) {
+  if ((!props.infinite && state.currentSlide === 0) || (props.infinite && state.totalItems < state.slidesToShow)) {
```
Comment (YIZHUANG) [inline src/utils/common.ts]: Try to reuse the utility function ```notEnoughChildren``` instead.

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG based on your comments; i revamped the code to only call setClones when we have enough children..  i also reversed the changes to the getClones made from earlier PR.. pls hold back merging.. as i am still confirming it will work with edge cases like resize, children changes etc.. 

I added a new resetTotalItems that will handle change in children better..

**Comment by YIZHUANG** [comment]
@abhinavdalal-iconnect 
I would actually prefer the changes you have 4 days ago, for example ````
function getClones(slidesToShow: number, childrenArr: any[]) {
  if (childrenArr.length < slidesToShow) {
    return childrenArr;
  }
}
```. 
I am not entire sure about the children changes yet, need to look into it too.

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG i have reverted; but kept some of the code back.. used notEnoughChildren wherever possible.. this is tested from my side now.. so if have a look at it and dont find any issues then you can merge it..

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG could you have a look at this as i need these changes released to then release my dependent code to prod..

**Comment by YIZHUANG** [comment]
Hi, I am too busy lately, I will need to test this when I have the time.

**Comment by abhinavdalal-iconnect** [comment]
@YIZHUANG  appreciate if you can have a look at this when you get the next opportunity as its currently a blocker for us.

File: src/Carousel.tsx [hunk 5]
```diff
@@ -611,11 +649,12 @@ class Carousel extends React.Component<CarouselProps, CarouselInternalState> {
     );
   }
   public renderCarouselItems() {
-    let clones = [];
-    if (this.props.infinite) {
-      const childrenArr = React.Children.toArray(this.props.children);
-      clones = getClones(this.state.slidesToShow, childrenArr);
-    }
+    const clones = this.props.infinite
```
Comment (YIZHUANG) [inline src/Carousel.tsx]: This is the same as the previous one, but this would get run anyway.

**Comment by abhinavdalal-iconnect** [comment]
closing this PR but will create 2 new ones

**Pull request closed without merging**